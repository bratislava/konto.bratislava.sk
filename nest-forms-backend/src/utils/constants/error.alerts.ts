import { FormError } from '@prisma/client'

import { ConvertErrorsEnum } from '../../convert/errors/convert.errors.enum'
import { FilesErrorsEnum } from '../../files/files.errors.enum'
import { NasesErrorsEnum } from '../../nases/nases.errors.enum'
import { ScannerClientErrorsEnum } from '../../scanner-client/scanner-client.errors.enum'
import { ErrorsEnum } from '../global-enums/errors.enum'

/**
 * This set contains all of the errors, which should be alerted in grafana when they are thrown.
 * If there is some new error enum, which should alert the developers, please add it to this list.
 * Do not add errors, which are not not necessary to alert, for example some NOT_FOUND errors from a controller,
 * invoked when a client tries to retrieve a form with nonexistent uuid.
 */
export default [
  ConvertErrorsEnum.ELEMENT_NOT_FOUND,
  ConvertErrorsEnum.UNPROCESSABLE_TYPE,
  FilesErrorsEnum.FILE_WRONG_STATUS_NOT_ACCEPTED_ERROR,
  FilesErrorsEnum.SCANNER_NO_RESPONSE_ERROR,
  FilesErrorsEnum.FILE_UPLOAD_TO_MINIO_WAS_NOT_SUCCESSFUL_ERROR,
  FilesErrorsEnum.FILE_DELETE_FROM_MINIO_WAS_NOT_SUCCESSFUL_ERROR,
  FilesErrorsEnum.FILE_DOWNLOAD_FROM_MINIO_WAS_NOT_SUCCESSFUL_ERROR,
  FilesErrorsEnum.FILE_MINIO_CHECK_ERROR,
  FilesErrorsEnum.FILE_BY_SCANNERID_NOT_FOUND_ERROR,
  NasesErrorsEnum.UNABLE_ADD_FORM_TO_RABBIT,
  NasesErrorsEnum.CITY_ACCOUNT_USER_GET_ERROR,
  NasesErrorsEnum.SEND_TO_NASES_ERROR,
  ScannerClientErrorsEnum.PROBLEM_WITH_SCANNER,
  ScannerClientErrorsEnum.FILE_HAS_WRONG_PARAMETERS,
  ScannerClientErrorsEnum.FILE_IN_SCANNER_NOT_FOUND,
  ErrorsEnum.DATABASE_ERROR,
  ErrorsEnum.INTERNAL_SERVER_ERROR,
  FormError.UNABLE_TO_SCAN_FILES,
  FormError.GINIS_SEND_ERROR,
  FormError.NASES_SEND_ERROR,
  FormError.RABBITMQ_MAX_TRIES,
] as string[]
