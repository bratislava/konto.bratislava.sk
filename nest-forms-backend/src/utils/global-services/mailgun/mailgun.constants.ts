/* eslint-disable no-secrets/no-secrets */
export enum MailgunTemplateEnum {
  NASES_SENT = 'NASES_SENT',
  GINIS_DELIVERED = 'GINIS_DELIVERED',
  NASES_GINIS_IN_PROGRESS = 'NASES_GINIS_IN_PROGRESS',
  GINIS_SUCCESS = 'GINIS_SUCCESS',
  GINIS_DENIED = 'GINIS_DENIED',
  ATTACHMENT_VIRUS = 'ATTACHMENT_VIRUS',
}

export enum MailgunConfigVariableType {
  PARAMETER = 'PARAMETER',
  SELECT = 'SELECT',
  STRING = 'STRING',
}

export const MAILGUN_CONFIG_FEEDBACK_URLS = {
  'zavazne-stanovisko-k-investicnej-cinnosti':
    'https://bravo.staffino.com/bratislava/id=WW1vhwT6',
  'stanovisko-k-investicnemu-zameru':
    'https://bravo.staffino.com/bratislava/id=WW1hkstR',
  'priznanie-k-dani-z-nehnutelnosti':
    'https://bravo.staffino.com/bratislava/id=WW14qo6q',
}

export const MAILGUN_CONFIG = {
  NASES_SENT: {
    template: '2023-application-status-sent',
    subject: 'Bratislavské konto: Vaša žiadosť bola odoslaná',
    variables: {
      applicationName: {
        type: MailgunConfigVariableType.PARAMETER,
        value: '{{messageSubject}}',
      },
      firstName: {
        type: MailgunConfigVariableType.PARAMETER,
        value: '{{firstName}}',
      },
      feHost: {
        type: MailgunConfigVariableType.STRING,
        value: process.env.FRONTEND_URL || 'https://konto.bratislava.sk',
      },
      feedbackLink: {
        type: MailgunConfigVariableType.SELECT,
        value: MAILGUN_CONFIG_FEEDBACK_URLS,
        selectorVariable: 'slug',
      },
    },
  },
  GINIS_DELIVERED: {
    template: '2023-application-status-delivered',
    subject: 'Bratislavské konto: Vaša žiadosť bola doručená',
    variables: {
      applicationName: {
        type: MailgunConfigVariableType.PARAMETER,
        value: '{{messageSubject}}',
      },
      firstName: {
        type: MailgunConfigVariableType.PARAMETER,
        value: '{{firstName}}',
      },
      feHost: {
        type: MailgunConfigVariableType.STRING,
        value: process.env.FRONTEND_URL || 'https://konto.bratislava.sk',
      },
      feedbackLink: {
        type: MailgunConfigVariableType.SELECT,
        value: MAILGUN_CONFIG_FEEDBACK_URLS,
        selectorVariable: 'slug',
      },
    },
  },
  NASES_GINIS_IN_PROGRESS: {
    template: '2023-application-status-in-progress',
    subject: 'Bratislavské konto: Na odoslaní Vašej žiadosti pracujeme',
    variables: {
      applicationName: {
        type: MailgunConfigVariableType.PARAMETER,
        value: '{{messageSubject}}',
      },
      firstName: {
        type: MailgunConfigVariableType.PARAMETER,
        value: '{{firstName}}',
      },
      feHost: {
        type: MailgunConfigVariableType.STRING,
        value: process.env.FRONTEND_URL || 'https://konto.bratislava.sk',
      },
    },
  },
  GINIS_SUCCESS: {
    template: '2023-application-status-success',
    subject: 'Bratislavské konto: Vaša žiadosť bola vybavená',
    variables: {
      applicationName: {
        type: MailgunConfigVariableType.PARAMETER,
        value: '{{messageSubject}}',
      },
      firstName: {
        type: MailgunConfigVariableType.PARAMETER,
        value: '{{firstName}}',
      },
      feHost: {
        type: MailgunConfigVariableType.STRING,
        value: process.env.FRONTEND_URL || 'https://konto.bratislava.sk',
      },
      feedbackLink: {
        type: MailgunConfigVariableType.SELECT,
        value: MAILGUN_CONFIG_FEEDBACK_URLS,
        selectorVariable: 'slug',
      },
    },
  },
  GINIS_DENIED: {
    template: '2023-application-status-denied',
    subject: 'Bratislavské konto: Vaša žiadosť bola zamietnutá',
    variables: {
      applicationName: {
        type: MailgunConfigVariableType.PARAMETER,
        value: '{{messageSubject}}',
      },
      firstName: {
        type: MailgunConfigVariableType.PARAMETER,
        value: '{{firstName}}',
      },
      feHost: {
        type: MailgunConfigVariableType.STRING,
        value: process.env.FRONTEND_URL || 'https://konto.bratislava.sk',
      },
      feedbackLink: {
        type: MailgunConfigVariableType.SELECT,
        value: MAILGUN_CONFIG_FEEDBACK_URLS,
        selectorVariable: 'slug',
      },
    },
  },
  ATTACHMENT_VIRUS: {
    template: '2023-application-status-virus',
    subject: 'Bratislavské konto: Vo Vašej žiadosti sa nachádza vírus',
    variables: {
      applicationName: {
        type: MailgunConfigVariableType.PARAMETER,
        value: '{{messageSubject}}',
      },
      firstName: {
        type: MailgunConfigVariableType.PARAMETER,
        value: '{{firstName}}',
      },
      feHost: {
        type: MailgunConfigVariableType.STRING,
        value: process.env.FRONTEND_URL || 'https://konto.bratislava.sk',
      },
      slug: {
        type: MailgunConfigVariableType.PARAMETER,
        value: `${
          process.env.FRONTEND_URL || 'https://konto.bratislava.sk'
        }/{{slug}}/{{formId}}`,
      },
    },
  },
}
/* eslint-enable no-secrets/no-secrets */
