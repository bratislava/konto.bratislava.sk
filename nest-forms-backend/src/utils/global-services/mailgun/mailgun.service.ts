import { Injectable } from '@nestjs/common'
import FormData from 'form-data'
import Mailgun from 'mailgun.js'
import { IMailgunClient } from 'mailgun.js/Interfaces'

// import { IMailgunClient } from 'mailgun.js/Interfaces'
import {
  SendEmailInputDto,
  SendEmailVariablesDto,
} from '../../global-dtos/mailgun.dto'
import {
  MAILGUN_CONFIG,
  MAILGUN_CONFIG_FEEDBACK_URLS,
  MailgunConfigVariableType,
} from './mailgun.constants'

@Injectable()
export default class MailgunService {
  mailgunClient: IMailgunClient

  constructor() {
    if (
      !process.env.MAILGUN_API_KEY ||
      !process.env.MAILGUN_HOST ||
      !process.env.MAILGUN_EMAIL_FROM ||
      !process.env.MAILGUN_DOMAIN
    ) {
      throw new Error(
        'Missing mailgun env, one of: MAILGUN_API_KEY, MAILGUN_HOST, MAILGUN_EMAIL_FROM, MAILGUN_DOMAIN',
      )
    }
    const mailgun = new Mailgun(FormData)
    this.mailgunClient = mailgun.client({
      username: 'api',
      key: process.env.MAILGUN_API_KEY || '',
      url: process.env.MAILGUN_HOST,
    })
  }

  private createEmailVariables(data: SendEmailInputDto): SendEmailVariablesDto {
    const response: SendEmailVariablesDto = {}
    Object.entries(MAILGUN_CONFIG[data.template].variables).forEach(
      ([key, val]) => {
        switch (val.type) {
          case MailgunConfigVariableType.PARAMETER:
            Object.entries(data.data).forEach(([k, v]) => {
              if (v) {
                response[key as keyof SendEmailVariablesDto] = (
                  response[key as keyof SendEmailVariablesDto] || val.value
                )
                  .toString()
                  .replace(`{{${k}}}`, v)
              }
            })

            break

          case MailgunConfigVariableType.SELECT:
            if (
              typeof val.value === 'object' &&
              Object.prototype.hasOwnProperty.call(val.value, data.data.slug)
            ) {
              response[key as keyof SendEmailVariablesDto] =
                val.value[
                  data.data.slug as keyof typeof MAILGUN_CONFIG_FEEDBACK_URLS
                ]
            }

            break

          case MailgunConfigVariableType.STRING:
            response[key as keyof SendEmailVariablesDto] = val.value
              ? val.value.toString()
              : ''

            break

          default:
            response[key as keyof SendEmailVariablesDto] = undefined
          // No default
        }
      },
    )
    return response
  }

  async sendEmail(data: SendEmailInputDto): Promise<void> {
    try {
      const mailgunResponse = await this.mailgunClient.messages.create(
        process.env.MAILGUN_DOMAIN!,
        {
          from: process.env.MAILGUN_EMAIL_FROM!,
          to: data.to,
          template: MAILGUN_CONFIG[data.template].template,
          subject: MAILGUN_CONFIG[data.template].subject,
          'h:X-Mailgun-Variables': JSON.stringify(
            this.createEmailVariables(data),
          ),
        },
      )
      if (mailgunResponse.status !== 200)
        console.warn(`Mailgun message was not sent for email `)
    } catch (error) {
      console.error('ERROR to send mailgun message', error)
    }
  }
}
