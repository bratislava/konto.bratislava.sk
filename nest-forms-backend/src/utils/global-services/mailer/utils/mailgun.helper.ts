import { Injectable } from '@nestjs/common'
import { ConfigService } from '@nestjs/config'
import FormData from 'form-data'
import Handlebars from 'handlebars'
import Mailgun, { TemplateQuery } from 'mailgun.js'
import { IMailgunClient } from 'mailgun.js/Interfaces'

import {
  SendEmailInputDto,
  SendEmailVariablesDto,
} from '../../../global-dtos/mailgun.dto'
import {
  MailgunErrorsEnum,
  MailgunErrorsResponseEnum,
} from '../../../global-enums/mailgun.errors.enum'
import ThrowerErrorGuard from '../../../guards/thrower-error.guard'
import { MAILGUN_CONFIG, MailgunConfigVariableType } from '../mailgun.constants'

@Injectable()
export default class MailgunHelper {
  mailgunClient: IMailgunClient

  constructor(
    private readonly throwerErrorGuard: ThrowerErrorGuard,
    private readonly configService: ConfigService,
  ) {
    if (
      !process.env.MAILGUN_API_KEY ||
      !process.env.MAILGUN_HOST ||
      !process.env.MAILGUN_EMAIL_FROM ||
      !process.env.MAILGUN_DOMAIN
    ) {
      throw new Error(
        'Missing mailgun env, one of: MAILGUN_API_KEY, MAILGUN_HOST, MAILGUN_EMAIL_FROM, MAILGUN_DOMAIN',
      )
    }

    const mailgun = new Mailgun(FormData)
    this.mailgunClient = mailgun.client({
      username: 'api',
      key: this.configService.getOrThrow('MAILGUN_API_KEY'),
      url: this.configService.getOrThrow('MAILGUN_HOST'),
    })
  }

  static createEmailVariables(data: SendEmailInputDto): SendEmailVariablesDto {
    const response: SendEmailVariablesDto = {}
    Object.entries(MAILGUN_CONFIG[data.template].variables).forEach(
      ([key, val]) => {
        switch (val.type) {
          case MailgunConfigVariableType.PARAMETER:
            Object.entries(data.data).forEach(([k, v]) => {
              if (v) {
                response[key as keyof SendEmailVariablesDto] = (
                  response[key as keyof SendEmailVariablesDto] || val.value
                )
                  .toString()
                  .replace(`{{${k}}}`, v)
              }
            })

            break

          case MailgunConfigVariableType.SELECT:
            if (
              typeof val.value === 'object' &&
              Object.prototype.hasOwnProperty.call(val.value, data.data.slug)
            ) {
              response[key as keyof SendEmailVariablesDto] =
                val.value[data.data.slug as keyof typeof val.value]
            }

            break

          case MailgunConfigVariableType.STRING:
            response[key as keyof SendEmailVariablesDto] = val.value
              ? val.value.toString()
              : ''

            break

          default:
            response[key as keyof SendEmailVariablesDto] = undefined
          // No default
        }
      },
    )
    return response
  }

  /**
   * Retrieves a Mailgun email template through API and fills it with provided variables using Handlebars.
   *
   * @async
   * @param {string} templateName - The name of the Mailgun template to retrieve.
   * @param {SendEmailVariablesDto} variables - An object containing variables to be inserted into the template.
   * @returns {Promise<string>} A promise that resolves to the filled template as an html string.
   * @throws {HttpException} Throws an error if the Mailgun template is not found.
   */
  async getFilledTemplate(
    templateName: string,
    variables: SendEmailVariablesDto,
  ): Promise<string> {
    const response = await this.mailgunClient.domains.domainTemplates.get(
      this.configService.getOrThrow('MAILGUN_DOMAIN'),
      templateName,
      { active: 'yes' } as TemplateQuery, // There is an error when importing mailgun.js/Enums, thus this is needed
    )

    if (!response || !response.version?.template) {
      throw this.throwerErrorGuard.NotFoundException(
        MailgunErrorsEnum.TEMPLATE_NOT_FOUND,
        `${MailgunErrorsResponseEnum.TEMPLATE_NOT_FOUND}: ${templateName}`,
      )
    }

    const compiledTemplate = Handlebars.compile(response.version.template)
    return compiledTemplate(variables)
  }
}
