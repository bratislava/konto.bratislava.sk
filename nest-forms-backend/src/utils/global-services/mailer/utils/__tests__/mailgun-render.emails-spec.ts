import { ConfigService } from '@nestjs/config'
import { Test, TestingModule } from '@nestjs/testing'

import ThrowerErrorGuard from '../../../../guards/thrower-error.guard'
import { MAILGUN_CONFIG } from '../../mailgun.constants'
import MailgunHelper from '../mailgun.helper'
import { testEmailDefinitions } from './utils/emailDefinitions'

describe('Mailgun Emails', () => {
  let mailgunHelper: MailgunHelper

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [MailgunHelper, ThrowerErrorGuard, ConfigService],
    }).compile()

    mailgunHelper = module.get<MailgunHelper>(MailgunHelper)
  })

  describe('local rendering of emails', () => {
    Object.entries(testEmailDefinitions).forEach(([key, emailDefinition]) => {
      it(`should correctly render email: ${key}`, async () => {
        expect(MAILGUN_CONFIG[emailDefinition.template].renderLocaly).toBe(true) // Ensure the email is set to be rendered locally

        const rendered = await mailgunHelper.getFilledTemplate(
          MAILGUN_CONFIG[emailDefinition.template].template,
          MailgunHelper.createEmailVariables(emailDefinition),
        )
        expect(rendered).toMatchSnapshot()
      })
    })
  })
})
