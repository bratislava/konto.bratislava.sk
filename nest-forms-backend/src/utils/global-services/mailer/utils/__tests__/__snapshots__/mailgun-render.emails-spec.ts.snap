// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Mailgun Emails local rendering of emails should correctly render email: test1 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html dir="ltr" lang="en">

  <head>
    <meta content="text/html; charset=UTF-8" http-equiv="Content-Type" />
    <meta name="x-apple-disable-message-reformatting" />
  </head>

  <body style="background-color:#f3f3f5;font-family:Inter, sans-serif;font-weight:400;font-size:16px;line-height:24px">
    <table align="center" width="100%" border="0" cellPadding="0" cellSpacing="0" role="presentation" style="max-width:584px;width:100%;margin:0 auto;background-color:#ffffff">
      <tbody>
        <tr style="width:100%">
          <td>
            <table align="center" width="100%" border="0" cellPadding="0" cellSpacing="0" role="presentation" style="border-bottom:solid 2px;border-color:#D6D6D6;padding-top:0;margin:0">
              <tbody style="width:100%">
                <tr style="width:100%">
                  <td data-id="__react-email-column">
                    <table width="100%" border="0" cellSpacing="0" cellPadding="0">
                      <tr>
                        <td align="center"><img alt="TSB logo" height="40" src="https://olo.s3.bratislava.sk/assets/email/tsb-logo.png" style="display:block;outline:none;border:none;text-decoration:none;padding:20px 0" /></td>
                      </tr>
                    </table>
                  </td>
                </tr>
              </tbody>
            </table>
            <table align="center" width="100%" border="0" cellPadding="0" cellSpacing="0" role="presentation" style="padding-bottom:32px;padding-left:24px;padding-right:24px">
              <tbody style="width:100%">
                <tr style="width:100%">
                  <td data-id="__react-email-column">
                    <table align="center" width="100%" border="0" cellPadding="0" cellSpacing="0" role="presentation">
                      <tbody style="width:100%">
                        <tr style="width:100%">
                          <td data-id="__react-email-column">
                            <div data-id="react-email-markdown">
                              <p style="margin-bottom:0px">tsb-form</p>
                            </div>
                          </td>
                        </tr>
                      </tbody>
                    </table>
                    <table align="center" width="100%" border="0" cellPadding="0" cellSpacing="0" role="presentation">
                      <tbody style="width:100%">
                        <tr style="width:100%">
                          <td data-id="__react-email-column"><p>This is a test submission.</p></td>
                        </tr>
                      </tbody>
                    </table>
                  </td>
                </tr>
              </tbody>
            </table>
          </td>
        </tr>
      </tbody>
    </table>
  </body>

</html>"
`;

exports[`Mailgun Emails local rendering of emails should correctly render email: test2 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html dir="ltr" lang="en">

  <head>
    <meta content="text/html; charset=UTF-8" http-equiv="Content-Type" />
    <meta name="x-apple-disable-message-reformatting" />
  </head>

  <body style="background-color:#f3f3f5;font-family:Inter, sans-serif;font-weight:400;font-size:16px;line-height:24px">
    <table align="center" width="100%" border="0" cellPadding="0" cellSpacing="0" role="presentation" style="max-width:584px;width:100%;margin:0 auto;background-color:#ffffff">
      <tbody>
        <tr style="width:100%">
          <td>
            <table align="center" width="100%" border="0" cellPadding="0" cellSpacing="0" role="presentation" style="border-bottom:solid 2px;border-color:#D6D6D6;padding-top:0;margin:0">
              <tbody style="width:100%">
                <tr style="width:100%">
                  <td data-id="__react-email-column">
                    <table width="100%" border="0" cellSpacing="0" cellPadding="0">
                      <tr>
                        <td align="center"><img alt="TSB logo" height="40" src="https://olo.s3.bratislava.sk/assets/email/tsb-logo.png" style="display:block;outline:none;border:none;text-decoration:none;padding:20px 0" /></td>
                      </tr>
                    </table>
                  </td>
                </tr>
              </tbody>
            </table>
            <table align="center" width="100%" border="0" cellPadding="0" cellSpacing="0" role="presentation" style="padding-bottom:32px;padding-left:24px;padding-right:24px">
              <tbody style="width:100%">
                <tr style="width:100%">
                  <td data-id="__react-email-column">
                    <table align="center" width="100%" border="0" cellPadding="0" cellSpacing="0" role="presentation">
                      <tbody style="width:100%">
                        <tr style="width:100%">
                          <td data-id="__react-email-column">
                            <div data-id="react-email-markdown">
                              <p style="margin-bottom:0px">tsb-form</p>
                            </div>
                          </td>
                        </tr>
                      </tbody>
                    </table>
                    <table align="center" width="100%" border="0" cellPadding="0" cellSpacing="0" role="presentation">
                      <tbody style="width:100%">
                        <tr style="width:100%">
                          <td data-id="__react-email-column">
      <table>
        <caption>Generated dataset â€” many rows and columns</caption>
        <thead>
          <tr>
            <th>#</th>
            <th>Alpha</th>
            <th>Bravo</th>
            <th>Charlie</th>
            <th>Delta</th>
            <th>Echo</th>
            <th>Foxtrot</th>
            <th>Golf</th>
            <th>Hotel</th>
            <th>India</th>
          </tr>
        </thead>
        <tbody>
          <!-- Rows start -->
          <!-- The following rows are intentionally repetitive; each cell contains a distinctive phrase so the total HTML size grows large. -->
          <!-- Row 1 -->
          <tr>
            <td>1</td>
            <td>Cell R1C2 - lorem ipsum dolor sit amet consectetur</td>
            <td>Cell R1C3 - consectetur adipiscing elit sed do eiusmod</td>
            <td>Cell R1C4 - tempor incididunt ut labore et dolore magna</td>
            <td>Cell R1C5 - aliqua ut enim ad minim veniam quis</td>
            <td>Cell R1C6 - nostrud exercitation ullamco laboris nisi ut</td>
            <td>Cell R1C7 - aliquip ex ea commodo consequat sunt in</td>
            <td>Cell R1C8 - culpa qui officia deserunt mollit anim id</td>
            <td>Cell R1C9 - est laborum repeated to enlarge the document</td>
            <td>Cell R1C10 - seed text for table growth and test use</td>
          </tr>
          <!-- Row 2 -->
          <tr>
            <td>2</td>
            <td>Cell R2C2 - sample phrase alpha beta gamma delta epsilon</td>
            <td>Cell R2C3 - sample phrase zeta eta theta iota kappa</td>
            <td>Cell R2C4 - repeated content to expand HTML size reliably</td>
            <td>Cell R2C5 - ensure this cell contains plenty of characters</td>
            <td>Cell R2C6 - plausible descriptive sentence to increase bytes</td>
            <td>Cell R2C7 - additional filler text to reach the requested size</td>
            <td>Cell R2C8 - more repeated words and phrases for test purposes</td>
            <td>Cell R2C9 - include links? no, just plain text for simplicity</td>
            <td>Cell R2C10 - stable filler content to make file suitably large</td>
          </tr>
          <!-- Row 3 through many -->
          <tr>
            <td>3</td>
            <td>Cell R3C2 - this is a deliberately long cell containing multiple words and phrases to grow the HTML</td>
            <td>Cell R3C3 - another long filler phrase used to increase the document length for testing</td>
            <td>Cell R3C4 - yet another repetitive but slightly varied sentence to avoid perfect duplication</td>
            <td>Cell R3C5 - testing the size boundary by including long human readable content</td>
            <td>Cell R3C6 - make sure the combined markup and content easily surpass sixteen kilobytes</td>
            <td>Cell R3C7 - keep adding sensible filler sentences so the HTML remains realistic</td>
            <td>Cell R3C8 - note this is not truly random but is varied enough for visual checking</td>
            <td>Cell R3C9 - content expanded with phrases, numbers, and small repeated clauses</td>
            <td>Cell R3C10 - end of row 3 filler content for table size testing</td>
          </tr>
          <tr>
            <td>4</td>
            <td>Cell R4C2 - continuing the pattern: data point, long sentence, extra details and notes</td>
            <td>Cell R4C3 - extend with additional descriptors to pad total HTML length significantly</td>
            <td>Cell R4C4 - filler text repeated in many variations to reach size threshold</td>
            <td>Cell R4C5 - content includes natural language that looks plausible for an email or report</td>
            <td>Cell R4C6 - more filler: this paragraph-like cell helps reach the 16 KB requirement</td>
            <td>Cell R4C7 - plenty of words here to ensure the combined payload grows fast</td>
            <td>Cell R4C8 - small sentences separated by commas to create moderate length</td>
            <td>Cell R4C9 - penultimate column for extra long repeated test content</td>
            <td>Cell R4C10 - final column of this row with additional descriptive filler</td>
          </tr>
          <tr>
            <td>5</td>
            <td>Cell R5C2 - sample long phrase A: lorem ipsum dolor sit amet, consetetur sadipscing elitr</td>
            <td>Cell R5C3 - sample long phrase B: sed diam nonumy eirmod tempor invidunt ut labore</td>
            <td>Cell R5C4 - sample long phrase C: et dolore magna aliquyam erat, sed diam voluptua</td>
            <td>Cell R5C5 - sample long phrase D: at vero eos et accusam et justo duo dolores</td>
            <td>Cell R5C6 - sample long phrase E: et ea rebum stet clita kasd gubergren no sea</td>
            <td>Cell R5C7 - sample long phrase F: takimata sanctus est Lorem ipsum dolor sit amet</td>
            <td>Cell R5C8 - sample long phrase G: consetetur sadipscing elitr sed diam nonumy</td>
            <td>Cell R5C9 - sample long phrase H: eirmod tempor invidunt ut labore et dolore magna</td>
            <td>Cell R5C10 - sample long phrase I: aliquyam erat sed diam voluptua vero eos</td>
          </tr>
          <tr>
            <td>6</td>
            <td>Cell R6C2 - auto-generated filler piece number six to add more content volume</td>
            <td>Cell R6C3 - auto-generated filler piece number six to add more content volume</td>
            <td>Cell R6C4 - auto-generated filler piece number six to add more content volume</td>
            <td>Cell R6C5 - auto-generated filler piece number six to add more content volume</td>
            <td>Cell R6C6 - auto-generated filler piece number six to add more content volume</td>
            <td>Cell R6C7 - auto-generated filler piece number six to add more content volume</td>
            <td>Cell R6C8 - auto-generated filler piece number six to add more content volume</td>
            <td>Cell R6C9 - auto-generated filler piece number six to add more content volume</td>
            <td>Cell R6C10 - auto-generated filler piece number six to add more content volume</td>
          </tr>
          <tr>
            <td>7</td>
            <td>Cell R7C2 - repeating varied sentence fragments: alpha, beta, gamma, delta, epsilon, zeta</td>
            <td>Cell R7C3 - repeating varied sentence fragments: eta, theta, iota, kappa, lambda, mu</td>
            <td>Cell R7C4 - repeating varied sentence fragments: nu, xi, omicron, pi, rho, sigma</td>
            <td>Cell R7C5 - repeating varied sentence fragments: tau, upsilon, phi, chi, psi, omega</td>
            <td>Cell R7C6 - padded content with commas and extra clauses for more bytes</td>
            <td>Cell R7C7 - padded content with commas and extra clauses for more bytes</td>
            <td>Cell R7C8 - padded content with commas and extra clauses for more bytes</td>
            <td>Cell R7C9 - padded content with commas and extra clauses for more bytes</td>
            <td>Cell R7C10 - padded content with commas and extra clauses for more bytes</td>
          </tr>
          <tr>
            <td>8</td>
            <td>Cell R8C2 - descriptive long sentence to simulate real-world email content or report block</td>
            <td>Cell R8C3 - descriptive long sentence to simulate real-world email content or report block</td>
            <td>Cell R8C4 - descriptive long sentence to simulate real-world email content or report block</td>
            <td>Cell R8C5 - descriptive long sentence to simulate real-world email content or report block</td>
            <td>Cell R8C6 - descriptive long sentence to simulate real-world email content or report block</td>
            <td>Cell R8C7 - descriptive long sentence to simulate real-world email content or report block</td>
            <td>Cell R8C8 - descriptive long sentence to simulate real-world email content or report block</td>
            <td>Cell R8C9 - descriptive long sentence to simulate real-world email content or report block</td>
            <td>Cell R8C10 - descriptive long sentence to simulate real-world email content or report block</td>
          </tr>
          <tr>
            <td>9</td>
            <td>Cell R9C2 - long repeated human readable chunk: test data, sample content, debug text</td>
            <td>Cell R9C3 - long repeated human readable chunk: test data, sample content, debug text</td>
            <td>Cell R9C4 - long repeated human readable chunk: test data, sample content, debug text</td>
            <td>Cell R9C5 - long repeated human readable chunk: test data, sample content, debug text</td>
            <td>Cell R9C6 - long repeated human readable chunk: test data, sample content, debug text</td>
            <td>Cell R9C7 - long repeated human readable chunk: test data, sample content, debug text</td>
            <td>Cell R9C8 - long repeated human readable chunk: test data, sample content, debug text</td>
            <td>Cell R9C9 - long repeated human readable chunk: test data, sample content, debug text</td>
            <td>Cell R9C10 - long repeated human readable chunk: test data, sample content, debug text</td>
          </tr>
          <tr>
            <td>10</td>
            <td>Cell R10C2 - continued filler: each of these cells contains dozens of characters to enlarge the file</td>
            <td>Cell R10C3 - continued filler: each of these cells contains dozens of characters to enlarge the file</td>
            <td>Cell R10C4 - continued filler: each of these cells contains dozens of characters to enlarge the file</td>
            <td>Cell R10C5 - continued filler: each of these cells contains dozens of characters to enlarge the file</td>
            <td>Cell R10C6 - continued filler: each of these cells contains dozens of characters to enlarge the file</td>
            <td>Cell R10C7 - continued filler: each of these cells contains dozens of characters to enlarge the file</td>
            <td>Cell R10C8 - continued filler: each of these cells contains dozens of characters to enlarge the file</td>
            <td>Cell R10C9 - continued filler: each of these cells contains dozens of characters to enlarge the file</td>
            <td>Cell R10C10 - continued filler: each of these cells contains dozens of characters to enlarge the file</td>
          </tr>
          <tr>
            <td>11</td>
            <td>Cell R11C2 - filler data to make document large enough for tests</td>
            <td>Cell R11C3 - filler data to make document large enough for tests</td>
            <td>Cell R11C4 - filler data to make document large enough for tests</td>
            <td>Cell R11C5 - filler data to make document large enough for tests</td>
            <td>Cell R11C6 - filler data to make document large enough for tests</td>
            <td>Cell R11C7 - filler data to make document large enough for tests</td>
            <td>Cell R11C8 - filler data to make document large enough for tests</td>
            <td>Cell R11C9 - filler data to make document large enough for tests</td>
            <td>Cell R11C10 - filler data to make document large enough for tests</td>
          </tr>
          <tr>
            <td>12</td>
            <td>Cell R12C2 - filler data to make document large enough for tests</td>
            <td>Cell R12C3 - filler data to make document large enough for tests</td>
            <td>Cell R12C4 - filler data to make document large enough for tests</td>
            <td>Cell R12C5 - filler data to make document large enough for tests</td>
            <td>Cell R12C6 - filler data to make document large enough for tests</td>
            <td>Cell R12C7 - filler data to make document large enough for tests</td>
            <td>Cell R12C8 - filler data to make document large enough for tests</td>
            <td>Cell R12C9 - filler data to make document large enough for tests</td>
            <td>Cell R12C10 - filler data to make document large enough for tests</td>
          </tr>
          <tr>
            <td>13</td>
            <td>Cell R13C2 - filler to expand size suitably for boundary testing</td>
            <td>Cell R13C3 - filler to expand size suitably for boundary testing</td>
            <td>Cell R13C4 - filler to expand size suitably for boundary testing</td>
            <td>Cell R13C5 - filler to expand size suitably for boundary testing</td>
            <td>Cell R13C6 - filler to expand size suitably for boundary testing</td>
            <td>Cell R13C7 - filler to expand size suitably for boundary testing</td>
            <td>Cell R13C8 - filler to expand size suitably for boundary testing</td>
            <td>Cell R13C9 - filler to expand size suitably for boundary testing</td>
            <td>Cell R13C10 - filler to expand size suitably for boundary testing</td>
          </tr>
          <tr>
            <td>14</td>
            <td>Cell R14C2 - extra long phrase, repeated for document inflation and test use</td>
            <td>Cell R14C3 - extra long phrase, repeated for document inflation and test use</td>
            <td>Cell R14C4 - extra long phrase, repeated for document inflation and test use</td>
            <td>Cell R14C5 - extra long phrase, repeated for document inflation and test use</td>
            <td>Cell R14C6 - extra long phrase, repeated for document inflation and test use</td>
            <td>Cell R14C7 - extra long phrase, repeated for document inflation and test use</td>
            <td>Cell R14C8 - extra long phrase, repeated for document inflation and test use</td>
            <td>Cell R14C9 - extra long phrase, repeated for document inflation and test use</td>
            <td>Cell R14C10 - extra long phrase, repeated for document inflation and test use</td>
          </tr>
        </tbody>
      </table>
      </td>
                        </tr>
                      </tbody>
                    </table>
                  </td>
                </tr>
              </tbody>
            </table>
          </td>
        </tr>
      </tbody>
    </table>
  </body>

</html>"
`;
